library/core/src/array/mod.rs

```diff
@@ -564,15 +525,97 @@ impl<T, const N: usize> [T; N] {
     /// ```
     #[unstable(feature = "array_methods", issue = "76118")]
     pub fn each_mut(&mut self) -> [&mut T; N] {
-        // Unlike in `map`, we don't need a guard here, as dropping a reference
-        // is a noop.
-        let mut out = MaybeUninit::uninit_array::<N>();
-        for (src, dst) in self.iter_mut().zip(&mut out) {
-            dst.write(src);
+        // SAFETY: we know for certain that this iterator will yield exactly `N`
+        // items.
+        unsafe { collect_into_array_unchecked(&mut self.iter_mut()) }
+    }
+}
+
+/// Pulls `N` items from `iter` and returns them as an array. If the iterator
+/// yields fewer than `N` items, this function exhibits undefined behavior.
+///
+/// If `iter.next()` panicks, all items already yielded by the iterator are
+/// dropped.
+///
+///
+/// # Safety
+///
+/// It is up to the caller to guarantee that `iter` yields at least `N` items.
+/// Violating this condition causes undefined behavior.
+unsafe fn collect_into_array_unchecked<I, const N: usize>(iter: &mut I) -> [I::Item; N]
+where
+    I: Iterator,
+{
+    match collect_into_array(iter) {
+        Some(array) => array,
+        // SAFETY: covered by the function contract.
+        None => unsafe { crate::hint::unreachable_unchecked() },
+    }
+}
+
+/// Pulls `N` items from `iter` and returns them as an array. If the iterator
+/// yields fewer than `N` items, `None` is returned and all already yielded
+/// items are dropped.
+///
+/// If `iter.next()` panicks, all items already yielded by the iterator are
+/// dropped.
+fn collect_into_array<I, const N: usize>(iter: &mut I) -> Option<[I::Item; N]>
+where
+    I: Iterator,
+{
+    if N == 0 {
+        // FIXME: is there a better way to achieve this?
+
+        // SAFETY: since N is 0, the array is empty and there are no
+        // uninitialized elements.
+        let out = unsafe { MaybeUninit::array_assume_init(MaybeUninit::uninit_array::<N>()) };
+        return Some(out);
```

Comment:
```suggestion
        return Some([]);
```

How about this?

Reply:
That's what I tried first. Unfortunately (but somewhat understandably) this results in:

```
error[E0308]: mismatched types
   --> library/core/src/array/mod.rs:569:21
    |
569 |         return Some([]);
    |                     ^^ expected `N`, found `0_usize`
    |
    = note: expected array `[<I as Iterator>::Item; N]`
               found array `[_; 0]`
```

Reply:
What if you do `return Some([] as [_; 0])`?

Reply:
That results in "non-primitive cast: `[_; 0]` as `[<I as Iterator>::Item; N]`". 

> I wonder why couldn't the compiler infer that `0_usize` type is `N`.

That would require control-flow based type analysis or something like that (a bit like Typescript, at least that's the only language I know that's comparable). The const is declared on the function and it's just `const N: usize`. In the `if` condition, it is just used as a value. If we would use specialization, and we would have an `impl` where `N` was set to 0 in the impl-header, then it might work.

Reply:
I think what you have is fine.

If you just want something shorter, it could be
```rust
// SAFETY: An empty array is always inhabited and has no validity invariants
return unsafe { Some(mem::zeroed()) };
```

Reply:
But in any case `Some([])` would be better. It will confuse newcomers to the language wondering why `[]` does not work as expected.

Reply:
I went with `mem::zeroed` now. Will resolve this thread (I usually don't resolve threads others opened, but I guess my own `FIXME` inspired this thread and I doubt any single one feels responsible for resolving.)